aten/src/ATen/core/ivalue.cpp

```diff
@@ -634,21 +634,47 @@ CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
   return ctx->dstFuture;
 }
 
-CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
-    std::vector<intrusive_ptr<ivalue::Future>> srcs) {
-  auto typePtr = srcs.empty() ? AnyType::get() : srcs[0]->elementType();
-  // Lists are generally expected to have a consistent type, check
-  // for this unless we have a compelling target use case.
-  for (size_t i = 1, len = srcs.size(); i < len; ++i) {
-    TORCH_CHECK(*srcs[i]->elementType() == *typePtr,
-                "Expected ", typePtr->str(), " type in list but saw ",
-                srcs[i]->elementType()->str());
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAny(
+    List<intrusive_ptr<ivalue::Future>> srcs) {
+  if (srcs.empty()) {
+    auto res = make_intrusive<ivalue::Future>(NoneType::get());
+    res->markCompleted();
+    return res;
   }
-  List<intrusive_ptr<ivalue::Future>> asList(FutureType::create(typePtr));
-  asList.reserve(srcs.size());
-  for (auto&& s : srcs) {
-    asList.push_back(std::move(s));
+  TypePtr typePtr = srcs.get(0)->elementType();
+  for (size_t i = 0, tot = srcs.size(); i < tot; ++i) {
+    if (srcs.get(i)->completed()) {
+      return srcs.get(i);
+    }
+    TORCH_CHECK(i == 0 || (*typePtr == *srcs.get(i)->elementType()));
```

Comment:
nit: we don't need i==0 here?

Reply:
The `i==0` is to skip the (more expensive) type check on the right for the i ==0 case